src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/AbstractTypeImportCompletionService.cs

```diff
@@ -372,15 +381,51 @@ public bool IsGeneric
             public bool IsAttribute
                 => (_properties & ItemPropertyKind.IsAttribute) != 0;
 
+            public bool IsEditorBrowsableStateAdvanced
+                => (_properties & ItemPropertyKind.IsEditorBrowsableStateAdvanced) != 0;
+
             public TypeImportCompletionItemInfo WithItem(CompletionItem item)
-                => new TypeImportCompletionItemInfo(item, IsPublic, IsGeneric, IsAttribute);
+                => new TypeImportCompletionItemInfo(item, IsPublic, IsGeneric, IsAttribute, IsEditorBrowsableStateAdvanced);
 
             [Flags]
             private enum ItemPropertyKind : byte
             {
                 IsPublic = 0x1,
                 IsGeneric = 0x2,
                 IsAttribute = 0x4,
+                IsEditorBrowsableStateAdvanced = 0x8,
+            }
+        }
+
+        // Things needed for determining whether a symbol is EditorBrowsable.
+        // Grouped together and reused within each compilation.
+        private class EditorBrowsableInfo
+        {
+            private Optional<IMethodSymbol?>? _editorBrowsableAttributeConstructor;
+            private ImmutableArray<IMethodSymbol>? _typeLibTypeAttributeConstructors;
+            private ImmutableArray<IMethodSymbol>? _typeLibFuncAttributeConstructors;
+            private ImmutableArray<IMethodSymbol>? _typeLibVarAttributeConstructors;
+
+            public Compilation Compilation { get; }
+
+            public Optional<IMethodSymbol?> EditorBrowsableAttributeConstructor
+                => _editorBrowsableAttributeConstructor ??= new(EditorBrowsableHelpers.GetSpecialEditorBrowsableAttributeConstructor(Compilation));
+
+            public ImmutableArray<IMethodSymbol> typeLibTypeAttributeConstructors
+                => _typeLibTypeAttributeConstructors ??= EditorBrowsableHelpers.GetSpecialTypeLibTypeAttributeConstructors(Compilation);
+
+            public ImmutableArray<IMethodSymbol> TypeLibFuncAttributeConstructors
+                => _typeLibFuncAttributeConstructors ??= EditorBrowsableHelpers.GetSpecialTypeLibFuncAttributeConstructors(Compilation);
+
+            public ImmutableArray<IMethodSymbol> TypeLibVarAttributeConstructors
+                => _typeLibVarAttributeConstructors ??= EditorBrowsableHelpers.GetSpecialTypeLibVarAttributeConstructors(Compilation);
+
+            public Optional<INamedTypeSymbol?> HideModuleNameAttribute { get; }
```

Comment:
why is this an optional nullable?

Reply:
it's not necessary, just so I can use the same `??=` operator to initialize it

Reply:
put is has no setter.  and is only initialized in the constructor.  so not sure why you need Optional at all...

Reply:
`Optional` is what eventually being passed as argument, so we don't need to create one at callsite.

Reply:
Also, user-defined conversion on Optional<T> is not allowed to convert from interface-types.